### ISSUE TITLE
Seurat2.3.4 not reproducing the same clusters for pbmc data set 

### ISSUE THREAD

**Issue by crazyhottommy** [satijalab/seurat#1058]
Hi Seurat developers,

I was trying to reproduce the same results following https://satijalab.org/seurat/pbmc3k_tutorial.html
I am copying the code and session info below

```r
library(Seurat)
library(dplyr)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "~/Downloads/filtered_gene_bc_matrices/hg19/")

pbmc <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 3, min.genes = 200, 
    project = "10X_PBMC")
mito.genes <- grep(pattern = "^MT-", x = rownames(x = pbmc@data), value = TRUE)
percent.mito <- Matrix::colSums(pbmc@raw.data[mito.genes, ])/Matrix::colSums(pbmc@raw.data)

# AddMetaData adds columns to object@meta.data, and is a great place to
# stash QC stats
pbmc <- AddMetaData(object = pbmc, metadata = percent.mito, col.name = "percent.mito")
VlnPlot(object = pbmc, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)

pbmc <- NormalizeData(object = pbmc, normalization.method = "LogNormalize", 
    scale.factor = 10000)

pbmc <- FindVariableGenes(object = pbmc, mean.function = ExpMean, dispersion.function = LogVMR, 
    x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)

length(x = pbmc@var.genes)

pbmc <- ScaleData(object = pbmc, vars.to.regress = c("nUMI", "percent.mito"))

pbmc <- RunPCA(object = pbmc, pc.genes = pbmc@var.genes, do.print = TRUE, pcs.print = 1:5, 
    genes.print = 5)

pbmc <- JackStraw(object = pbmc, num.replicate = 100, display.progress = FALSE)

JackStrawPlot(object = pbmc, PCs = 1:12)

pbmc <- FindClusters(object = pbmc, reduction.type = "pca", dims.use = 1:9, 
    resolution = 0.6, print.output = 0, save.SNN = TRUE)

pbmc <- RunTSNE(object = pbmc, dims.use = 1:10, do.fast = TRUE)

TSNEPlot(object = pbmc, do.label = T)

```
![pbmc_test_tnse](https://user-images.githubusercontent.com/4106146/51138809-1dcc9180-1810-11e9-95a0-5e05e038b84e.png)

compare it with the one in the tutorial, the 0 cluster is splitted into 0 and 1. and the small cluster Megakaryocytes is gone.

Anything changed?
```
devtools::session_info()
─ Session info ──────────────────────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.5.1 (2018-07-02)
 os       macOS High Sierra 10.13.6   
 system   x86_64, darwin15.6.0        
 ui       RStudio                     
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/New_York            
 date     2019-01-14                  

─ Packages ──────────────────────────────────────────────────────────────────────────────────────────
 package              * version    date       lib source                             
 acepack                1.4.1      2016-10-29 [1] CRAN (R 3.5.0)                     
 ape                    5.2        2018-09-24 [1] CRAN (R 3.5.0)                     
 assertthat             0.2.0      2017-04-11 [1] CRAN (R 3.5.0)                     
 backports              1.1.3      2018-12-14 [1] CRAN (R 3.5.0)                     
 base64enc              0.1-3      2015-07-28 [1] CRAN (R 3.5.0)                     
 beeswarm               0.2.3      2016-04-25 [1] CRAN (R 3.5.0)                     
 bibtex                 0.4.2      2017-06-30 [1] CRAN (R 3.5.0)                     
 bindr                  0.1.1      2018-03-13 [1] CRAN (R 3.5.0)                     
 bindrcpp             * 0.2.2      2018-03-29 [1] CRAN (R 3.5.0)                     
 Biobase              * 2.42.0     2018-10-30 [1] Bioconductor                       
 BiocGenerics         * 0.28.0     2018-10-30 [1] Bioconductor                       
 BiocManager            1.30.4     2018-11-13 [1] CRAN (R 3.5.0)                     
 BiocNeighbors          1.0.0      2018-10-30 [1] Bioconductor                       
 BiocParallel         * 1.16.2     2018-11-28 [1] Bioconductor                       
 bit                    1.1-14     2018-05-29 [1] CRAN (R 3.5.0)                     
 bit64                  0.9-7      2017-05-08 [1] CRAN (R 3.5.0)                     
 bitops                 1.0-6      2013-08-17 [1] CRAN (R 3.5.0)                     
 broom                  0.5.1      2018-12-05 [1] CRAN (R 3.5.0)                     
 callr                  3.1.1      2018-12-21 [1] CRAN (R 3.5.0)                     
 caTools                1.17.1.1   2018-07-20 [1] CRAN (R 3.5.0)                     
 cellranger             1.1.0      2016-07-27 [1] CRAN (R 3.5.0)                     
 checkmate              1.8.5      2017-10-24 [1] CRAN (R 3.5.0)                     
 circlize               0.4.5      2018-11-21 [1] CRAN (R 3.5.0)                     
 class                  7.3-15     2019-01-01 [1] CRAN (R 3.5.2)                     
 cli                    1.0.1      2018-09-25 [1] CRAN (R 3.5.0)                     
 cluster                2.0.7-1    2018-04-13 [1] CRAN (R 3.5.1)                     
 codetools              0.2-16     2018-12-24 [1] CRAN (R 3.5.2)                     
 colorspace             1.3-2      2016-12-14 [1] CRAN (R 3.5.0)                     
 ComplexHeatmap         1.20.0     2018-10-30 [1] Bioconductor                       
 cowplot              * 0.9.3      2018-07-15 [1] CRAN (R 3.5.0)                     
 crayon                 1.3.4      2017-09-16 [1] CRAN (R 3.5.0)                     
 data.table             1.11.8     2018-09-30 [1] CRAN (R 3.5.0)                     
 DelayedArray         * 0.8.0      2018-10-30 [1] Bioconductor                       
 DelayedMatrixStats     1.4.0      2018-10-30 [1] Bioconductor                       
 DEoptimR               1.0-8      2016-11-19 [1] CRAN (R 3.5.0)                     
 desc                   1.2.0      2018-05-01 [1] CRAN (R 3.5.0)                     
 devtools               2.0.1      2018-10-26 [1] CRAN (R 3.5.1)                     
 digest                 0.6.18     2018-10-10 [1] CRAN (R 3.5.0)                     
 diptest                0.75-7     2016-12-05 [1] CRAN (R 3.5.0)                     
 doSNOW                 1.0.16     2017-12-13 [1] CRAN (R 3.5.0)                     
 dplyr                * 0.7.8      2018-11-10 [1] CRAN (R 3.5.0)                     
 dtw                    1.20-1     2018-05-18 [1] CRAN (R 3.5.0)                     
 dynamicTreeCut         1.63-1     2016-03-11 [1] CRAN (R 3.5.0)                     
 edgeR                  3.24.2     2018-12-16 [1] Bioconductor                       
 evaluate               0.12       2018-10-09 [1] CRAN (R 3.5.0)                     
 fansi                  0.4.0      2018-10-05 [1] CRAN (R 3.5.0)                     
 fitdistrplus           1.0-11     2018-09-10 [1] CRAN (R 3.5.0)                     
 flexmix                2.3-14     2017-04-28 [1] CRAN (R 3.5.0)                     
 forcats              * 0.3.0      2018-02-19 [1] CRAN (R 3.5.0)                     
 foreach                1.4.4      2017-12-12 [1] CRAN (R 3.5.0)                     
 foreign                0.8-71     2018-07-20 [1] CRAN (R 3.5.0)                     
 Formula                1.2-3      2018-05-03 [1] CRAN (R 3.5.0)                     
 fpc                    2.1-11.1   2018-07-20 [1] CRAN (R 3.5.0)                     
 fs                     1.2.6      2018-08-23 [1] CRAN (R 3.5.0)                     
 gbRd                   0.4-11     2012-10-01 [1] CRAN (R 3.5.0)                     
 gdata                  2.18.0     2017-06-06 [1] CRAN (R 3.5.0)                     
 generics               0.0.2      2018-11-29 [1] CRAN (R 3.5.0)                     
 GenomeInfoDb         * 1.18.1     2018-11-12 [1] Bioconductor                       
 GenomeInfoDbData       1.2.0      2019-01-02 [1] Bioconductor                       
 GenomicRanges        * 1.34.0     2018-10-30 [1] Bioconductor                       
 GetoptLong             0.1.7      2018-06-10 [1] CRAN (R 3.5.0)                     
 ggbeeswarm             0.6.0      2017-08-07 [1] CRAN (R 3.5.0)                     
 ggimage                0.2.0      2018-12-10 [1] CRAN (R 3.5.0)                     
 ggplot2              * 3.1.0      2018-10-25 [1] CRAN (R 3.5.0)                     
 ggplotify              0.0.3      2018-08-03 [1] CRAN (R 3.5.0)                     
 ggrepel              * 0.8.0      2018-05-09 [1] CRAN (R 3.5.0)                     
 ggridges               0.5.1      2018-09-27 [1] CRAN (R 3.5.0)                     
 GlobalOptions          0.1.0      2018-06-09 [1] CRAN (R 3.5.0)                     
 glue                   1.3.0      2018-07-17 [1] CRAN (R 3.5.0)                     
 gplots                 3.0.1      2016-03-30 [1] CRAN (R 3.5.0)                     
 gridExtra              2.3        2017-09-09 [1] CRAN (R 3.5.0)                     
 gridGraphics           0.3-0      2018-04-03 [1] CRAN (R 3.5.0)                     
 gtable                 0.2.0      2016-02-26 [1] CRAN (R 3.5.0)                     
 gtools                 3.8.1      2018-06-26 [1] CRAN (R 3.5.0)                     
 haven                  2.0.0      2018-11-22 [1] CRAN (R 3.5.0)                     
 HDF5Array              1.10.1     2018-12-05 [1] Bioconductor                       
 hdf5r                  1.0.1      2018-10-07 [1] CRAN (R 3.5.0)                     
 here                 * 0.1        2017-05-28 [1] CRAN (R 3.5.0)                     
 hexbin                 1.27.2     2018-01-15 [1] CRAN (R 3.5.0)                     
 hexSticker           * 0.4.4      2018-06-22 [1] CRAN (R 3.5.0)                     
 Hmisc                  4.1-1      2018-01-03 [1] CRAN (R 3.5.0)                     
 hms                    0.4.2      2018-03-10 [1] CRAN (R 3.5.0)                     
 htmlTable              1.12       2018-05-26 [1] CRAN (R 3.5.0)                     
 htmltools              0.3.6      2017-04-28 [1] CRAN (R 3.5.0)                     
 htmlwidgets            1.3        2018-09-30 [1] CRAN (R 3.5.0)                     
 httr                   1.4.0      2018-12-11 [1] CRAN (R 3.5.0)                     
 ica                    1.0-2      2018-05-24 [1] CRAN (R 3.5.0)                     
 igraph                 1.2.2      2018-07-27 [1] CRAN (R 3.5.0)                     
 IRanges              * 2.16.0     2018-10-30 [1] Bioconductor                       
 irlba                  2.3.2      2018-01-11 [1] CRAN (R 3.5.0)                     
 iterators              1.0.10     2018-07-13 [1] CRAN (R 3.5.0)                     
 jsonlite               1.6        2018-12-07 [1] CRAN (R 3.5.0)                     
 kernlab                0.9-27     2018-08-10 [1] CRAN (R 3.5.0)                     
 KernSmooth             2.23-15    2015-06-29 [1] CRAN (R 3.5.1)                     
 knitr                  1.21       2018-12-10 [1] CRAN (R 3.5.1)                     
 labeling               0.3        2014-08-23 [1] CRAN (R 3.5.0)                     
 lars                   1.2        2013-04-24 [1] CRAN (R 3.5.0)                     
 lattice                0.20-38    2018-11-04 [1] CRAN (R 3.5.0)                     
 latticeExtra           0.6-28     2016-02-09 [1] CRAN (R 3.5.0)                     
 lazyeval               0.2.1      2017-10-29 [1] CRAN (R 3.5.0)                     
 limma                  3.38.3     2018-12-02 [1] Bioconductor                       
 lmtest                 0.9-36     2018-04-04 [1] CRAN (R 3.5.0)                     
 locfit                 1.5-9.1    2013-04-20 [1] CRAN (R 3.5.0)                     
 lsei                   1.2-0      2017-10-23 [1] CRAN (R 3.5.0)                     
 lubridate              1.7.4      2018-04-11 [1] CRAN (R 3.5.0)                     
 magick                 2.0        2018-10-05 [1] CRAN (R 3.5.0)                     
 magrittr               1.5        2014-11-22 [1] CRAN (R 3.5.0)                     
 MASS                   7.3-51.1   2018-11-01 [1] CRAN (R 3.5.0)                     
 Matrix               * 1.2-15     2018-11-01 [1] CRAN (R 3.5.0)                     
 matrixStats          * 0.54.0     2018-07-23 [1] CRAN (R 3.5.0)                     
 mclust                 5.4.2      2018-11-17 [1] CRAN (R 3.5.0)                     
 memoise                1.1.0      2017-04-21 [1] CRAN (R 3.5.0)                     
 metap                  1.0        2018-07-25 [1] CRAN (R 3.5.0)                     
 mixtools               1.1.0      2017-03-10 [1] CRAN (R 3.5.0)                     
 modelr                 0.1.2      2018-05-11 [1] CRAN (R 3.5.0)                     
 modeltools             0.2-22     2018-07-16 [1] CRAN (R 3.5.0)                     
 munsell                0.5.0      2018-06-12 [1] CRAN (R 3.5.0)                     
 mvtnorm                1.0-8      2018-05-31 [1] CRAN (R 3.5.0)                     
 nlme                   3.1-137    2018-04-07 [1] CRAN (R 3.5.1)                     
 nnet                   7.3-12     2016-02-02 [1] CRAN (R 3.5.1)                     
 npsurv                 0.4-0      2017-10-14 [1] CRAN (R 3.5.0)                     
 patchwork            * 0.0.1      2018-10-02 [1] Github (dynverse/patchwork@c518d36)
 pbapply                1.3-4      2018-01-10 [1] CRAN (R 3.5.0)                     
 pillar                 1.3.1      2018-12-15 [1] CRAN (R 3.5.0)                     
 pkgbuild               1.0.2      2018-10-16 [1] CRAN (R 3.5.0)                     
 pkgconfig              2.0.2      2018-08-16 [1] CRAN (R 3.5.0)                     
 pkgload                1.0.2      2018-10-29 [1] CRAN (R 3.5.0)                     
 plyr                   1.8.4      2016-06-08 [1] CRAN (R 3.5.0)                     
 png                    0.1-7      2013-12-03 [1] CRAN (R 3.5.0)                     
 prabclus               2.2-6      2015-01-14 [1] CRAN (R 3.5.0)                     
 prettyunits            1.0.2      2015-07-13 [1] CRAN (R 3.5.0)                     
 processx               3.2.1      2018-12-05 [1] CRAN (R 3.5.0)                     
 proxy                  0.4-22     2018-04-08 [1] CRAN (R 3.5.0)                     
 ps                     1.3.0      2018-12-21 [1] CRAN (R 3.5.0)                     
 purrr                * 0.2.5      2018-05-29 [1] CRAN (R 3.5.0)                     
 R.methodsS3            1.7.1      2016-02-16 [1] CRAN (R 3.5.0)                     
 R.oo                   1.22.0     2018-04-22 [1] CRAN (R 3.5.0)                     
 R.utils                2.7.0      2018-08-27 [1] CRAN (R 3.5.0)                     
 R6                     2.3.0      2018-10-04 [1] CRAN (R 3.5.0)                     
 RANN                   2.6        2018-07-16 [1] CRAN (R 3.5.0)                     
 RColorBrewer           1.1-2      2014-12-07 [1] CRAN (R 3.5.0)                     
 Rcpp                   1.0.0      2018-11-07 [1] CRAN (R 3.5.0)                     
 RCurl                  1.95-4.11  2018-07-15 [1] CRAN (R 3.5.0)                     
 Rdpack                 0.10-1     2018-10-04 [1] CRAN (R 3.5.0)                     
 readr                * 1.3.1      2018-12-21 [1] CRAN (R 3.5.0)                     
 readxl                 1.2.0      2018-12-19 [1] CRAN (R 3.5.0)                     
 remotes                2.0.2      2018-10-30 [1] CRAN (R 3.5.0)                     
 reshape2               1.4.3      2017-12-11 [1] CRAN (R 3.5.0)                     
 reticulate             1.10       2018-08-05 [1] CRAN (R 3.5.0)                     
 rhdf5                  2.26.1     2018-12-18 [1] Bioconductor                       
 Rhdf5lib               1.4.2      2018-12-03 [1] Bioconductor                       
 rjson                  0.2.20     2018-06-08 [1] CRAN (R 3.5.0)                     
 rlang                  0.3.0.1    2018-10-25 [1] CRAN (R 3.5.0)                     
 rmarkdown              1.11       2018-12-08 [1] CRAN (R 3.5.0)                     
 robustbase             0.93-3     2018-09-21 [1] CRAN (R 3.5.0)                     
 ROCR                   1.0-7      2015-03-26 [1] CRAN (R 3.5.0)                     
 rpart                  4.1-13     2018-02-23 [1] CRAN (R 3.5.1)                     
 rprojroot              1.3-2      2018-01-03 [1] CRAN (R 3.5.0)                     
 rstudioapi             0.8        2018-10-02 [1] CRAN (R 3.5.0)                     
 Rtsne                  0.15       2018-11-10 [1] CRAN (R 3.5.0)                     
 rvcheck                0.1.3      2018-12-06 [1] CRAN (R 3.5.0)                     
 rvest                  0.3.2      2016-06-17 [1] CRAN (R 3.5.0)                     
 S4Vectors            * 0.20.1     2018-11-09 [1] Bioconductor                       
 scales                 1.0.0      2018-08-09 [1] CRAN (R 3.5.0)                     
 scater               * 1.10.0     2018-10-30 [1] Bioconductor                       
 scclusteval          * 0.0.0.9000 2019-01-11 [1] local                              
 scran                * 1.10.1     2018-11-03 [1] Bioconductor                       
 SDMTools               1.1-221    2014-08-05 [1] CRAN (R 3.5.0)                     
 segmented              0.5-3.0    2017-11-30 [1] CRAN (R 3.5.0)                     
 sessioninfo            1.1.1      2018-11-05 [1] CRAN (R 3.5.0)                     
 Seurat               * 2.3.4      2018-07-17 [1] CRAN (R 3.5.0)                     
 shape                  1.4.4      2018-02-07 [1] CRAN (R 3.5.0)                     
 showtext               0.5-1      2018-01-16 [1] CRAN (R 3.5.0)                     
 showtextdb             2.0        2017-09-11 [1] CRAN (R 3.5.0)                     
 SingleCellExperiment * 1.4.0      2018-10-30 [1] Bioconductor                       
 snow                   0.4-3      2018-09-14 [1] CRAN (R 3.5.0)                     
 splatter             * 1.6.1      2018-12-06 [1] Bioconductor                       
 statmod                1.4.30     2017-06-18 [1] CRAN (R 3.5.0)                     
 stringi                1.2.4      2018-07-20 [1] CRAN (R 3.5.0)                     
 stringr              * 1.3.1      2018-05-10 [1] CRAN (R 3.5.0)                     
 SummarizedExperiment * 1.12.0     2018-10-30 [1] Bioconductor                       
 survival               2.43-3     2018-11-26 [1] CRAN (R 3.5.0)                     
 sysfonts               0.8        2018-10-11 [1] CRAN (R 3.5.0)                     
 testthat               2.0.1      2018-10-13 [1] CRAN (R 3.5.0)                     
 tibble               * 1.4.2      2018-01-22 [1] CRAN (R 3.5.0)                     
 tidyr                * 0.8.2      2018-10-28 [1] CRAN (R 3.5.0)                     
 tidyselect             0.2.5      2018-10-11 [1] CRAN (R 3.5.0)                     
 tidyverse            * 1.2.1      2017-11-14 [1] CRAN (R 3.5.0)                     
 trimcluster            0.1-2.1    2018-07-20 [1] CRAN (R 3.5.0)                     
 tsne                   0.1-3      2016-07-15 [1] CRAN (R 3.5.0)                     
 usethis                1.4.0      2018-08-14 [1] CRAN (R 3.5.0)                     
 utf8                   1.1.4      2018-05-24 [1] CRAN (R 3.5.0)                     
 vipor                  0.4.5      2017-03-22 [1] CRAN (R 3.5.0)                     
 viridis                0.5.1      2018-03-29 [1] CRAN (R 3.5.0)                     
 viridisLite            0.3.0      2018-02-01 [1] CRAN (R 3.5.0)                     
 withr                  2.1.2      2018-03-15 [1] CRAN (R 3.5.0)                     
 workflowr            * 1.1.1      2018-07-06 [1] CRAN (R 3.5.0)                     
 xfun                   0.4        2018-10-23 [1] CRAN (R 3.5.0)                     
 xml2                   1.2.0      2018-01-24 [1] CRAN (R 3.5.0)                     
 XVector                0.22.0     2018-10-30 [1] Bioconductor                       
 yaml                   2.2.0      2018-07-25 [1] CRAN (R 3.5.0)                     
 zlibbioc               1.28.0     2018-10-30 [1] Bioconductor                       
 zoo                    1.8-4      2018-09-19 [1] CRAN (R 3.5.0)                     

[1] /Library/Frameworks/R.framework/Versions/3.5/Resources/library
```

**Comment by timoast** [comment]
Here you're using the first 9 PCs for clustering, whereas in the tutorial we use the first 10:

```r
pbmc <- FindClusters(object = pbmc, reduction.type = "pca", dims.use = 1:9, 
    resolution = 0.6, print.output = 0, save.SNN = TRUE)
```

vs

```r
pbmc <- FindClusters(object = pbmc, reduction.type = "pca", dims.use = 1:10, 
    resolution = 0.6, print.output = 0, save.SNN = TRUE)
```

**Comment by crazyhottommy** [comment]
Thanks for the quick answer. I tried dim.use = 1:10 as well, indeed, it reproduced the same number of clusters. So the number of PCs used has a great effect on the final clusters. according to the JackStrawPlot output, PC10 has a p-value of 0.479 (this is different from the tutorial and expected since the 100 runs are random), that's why I used PC from 1:9 for finding clusters.

What's the best practice in choosing the number of PCs then?
